ARROW-15538: [C++] Expanding coverage of math functions from Substrait to Acero

drabastomek opened the pull request:
This PR adds the following functionality to Acero/Substrait consumer

1. Support for logarithmic functions.
2. Support for power, sqrt, exp, abs, sign.
3. Implements exp kernel.

drabastomek commented:
Thanks @jvanstraten. You're right about the `exp` not being defined for ints - I changed the calls and added `_checked` method too. 

I agree that currently all the `_checked` methods only check for int overflows and it might be something we might want to address at some point. Not sure if this should be a part of this PR tho.

westonpace commented on cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:
```diff
@@ -1717,6 +1717,12 @@ const FunctionDoc pow_doc{
     {"base", "exponent"}};
 
+const FunctionDoc exp_doc{
+    "Computes Euler's number raised to the specified power, element-wise",
```
```suggestion
    "Compute Euler's number raised to the specified power, element-wise",
```

drabastomek replied:
Corrected the `exp_doc`.

westonpace commented on cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:
```diff
@@ -1717,6 +1717,12 @@ const FunctionDoc pow_doc{
 
+const FunctionDoc exp_doc{
+    "Computes Euler's number raised to the specified power, element-wise",
+    ("Negative integer power returns an error. However, integer overflow\n"
```
```suggestion
    ("Negative integer power returns an error. However, integer overflow\n"
```
What does negative integer power mean?  This function only has non-integral kernels.

I'm not sure what you mean by integer overflow.  Why "integer"?  The return value is a floating point.  If there is an overflow then "[+HUGE_VAL](https://en.cppreference.com/w/cpp/numeric/math/HUGE_VAL), +HUGE_VALF, or +HUGE_VALL is returned."

drabastomek replied:
That was a placeholder from power. I updated it.

westonpace commented on cpp/src/arrow/engine/substrait/extension_set.cc:
```diff
@@ -858,11 +858,27 @@ struct DefaultExtensionIdRegistry : ExtensionIdRegistryImpl {
+    for (const auto& function_name : {"ceil", "floor"}) {
+      DCHECK_OK(
+          AddSubstraitCallToArrow({kSubstraitRoundingFunctionsUri, function_name},
+                                  DecodeOptionlessOverflowableArithmetic(function_name)));
```
Why use "Overflowable" here?  I don't think there is a `ceil_checked` or `floor_checked`?

drabastomek replied:
I noticed that after. Updated to `DecodeOptionlessUncheckedArithmetic`.

westonpace commented:
> I don't think DecodeOptionlessOverflowableArithmetic is suitable for these functions, nor for division or for any of the floating-point versions of the functions it's already used for. 

At the moment we should focus on exposing the capability that we have, and describing it accurately.  For example, in Arrow we have "_checked" but "what exactly is checked?"  For example, in Arrow we have `divide` and `divide_checked`.  They seem to behave as follows (through trial and error):

| Method | Data Type | Problem | Behavior |
| --- | --- | --- | --- |
| divide | integral | overflow | silent |
| divide | integral | divide by zero | error |
| divide | integral | domain error | error |
| divide | floating | divide by zero | limit |
| divide | floating | domain error | nan |
| divide_checked | integral | overflow | error |
| divide_checked | integral | divide by zero | error |
| divide_checked | integral | domain error | error |
| divide_checked | floating | divide by zero | error |
| divide_checked | floating | domain error | error |

Thus we should have the following behavior for substrait calls

| Call | Data Type | Options | Arrow Result |
| --- | --- | --- | --- |
| divide | integral | overflow=SILENT | divide |
| divide | integral | overflow=SATURATE | reject plan |
| divide | integral | overflow=ERROR | divide_checked |
| divide | floating | rounding=* | reject plan[1] |
| divide | floating | on_domain_error=NAN | divide |
| divide | floating | on_domain_error=ERROR | divide_checked |
| divide | floating | on_division_by_zero=LIMIT | divide |
| divide | floating | on_division_by_zero=NAN | reject plan |
| divide | floating | on_division_by_zero=ERROR | divide_checked |
| divide | floating | on_domain_error=NAN on_division_by_zero=ERROR | reject plan[2] |
| divide | floating | on_domain_error=ERROR on_division_by_zero=LIMIT | reject plan[2] |

1. I'm not sure what rounding behavior we have implemented in Acero and therefore the safest thing to do would be to reject all plans that are asking for a specific rounding behavior.
2. We cannot satisfy these specific combinations between divide and divide_checked.

Caveat / Needs clarification:

IMHO, It is not clear, in Substrait, if division by zero for integral arguments counts as overflow or not.  Also, it is possible to have 0/0 in integer arithmetic, which I have been interpreting as "domain error" (technically division by zero requires non-zero/0) but have no idea what the integral behavior should be.

drabastomek commented on cpp/src/arrow/compute/kernels/base_arithmetic_internal.h:
```diff
@@ -485,6 +485,34 @@ struct NegateChecked {
+  }
+};
+
+struct ExpChecked {
```
Good point. Removed the checks for `exp` and added `DecodeOptionlessUncheckedMapping` method to handle these cases.

jvanstraten commented:
> I'm not sure what rounding behavior we have implemented in Acero and therefore the safest thing to do would be to reject all plans that are asking for a specific rounding behavior.

Assuming that the arithmetic operations are all ultimately implemented using standard library operations, you can query the current mode with [`fegetround()`](https://en.cppreference.com/w/cpp/numeric/fenv/feround) (it's a global processor state thing). But unless an Arrow user goes out of their way to change the mode and this somehow propagates to all threads correctly, it's just going to be `FE_TONEAREST`, which corresponds with Substrait `TIE_TO_EVEN` (`TIE_AWAY_FROM_ZERO` does not seem to be implemented by the standard library, the rest of the mapping should be fairly straightforward).

westonpace commented:
> It seems like a safe bet to me to accept both unspecified and TIE_TO_EVEN.

Worst case we find out when we get around to implementing Substrait function testing for this specific option :laughing: Nightmare fuel for @richtia

drabastomek commented on cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:
```diff
@@ -2224,6 +2230,15 @@ void RegisterScalarArithmetic(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(power_checked)));
 
+  // ----------------------------------------------------------------------
+  auto exp = MakeUnaryArithmeticFunctionFloatingPoint<Exp>("exp", exp_doc);
```
Tests added.

westonpace commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -177,6 +177,31 @@ TEST(FunctionMapping, ValidCases) {
        "0",
        int8()},
+      {{kSubstraitArithmeticFunctionsUri, "sign"},
+       {"SILENT", "-1"},
```
```suggestion
       {"-1"},
```
`sign` does not have an overflow argument.

drabastomek replied:
Changed. Thanks for catching this.

westonpace commented on cpp/src/arrow/engine/substrait/extension_set.cc:
```diff
@@ -858,11 +867,34 @@ struct DefaultExtensionIdRegistry : ExtensionIdRegistryImpl {
     // Mappings with a _checked variant
-    for (const auto& function_name : {"add", "subtract", "multiply", "divide"}) {
+    for (const auto& function_name :
+         {"add", "subtract", "multiply", "divide", "power", "sqrt", "abs"}) {
```
`sqrt` isn't currently defined as overflowable in Substrait.

drabastomek replied:
Ha -- good catch. I changed the `DecodeOptionlessUncheckedArithmetic` and updated the tests.

westonpace commented on cpp/src/arrow/engine/substrait/extension_set.cc:
```diff
@@ -858,11 +867,34 @@ struct DefaultExtensionIdRegistry : ExtensionIdRegistryImpl {
+    for (const auto& function_name : {"ln", "log10", "log2", "logb", "log1p"}) {
+      DCHECK_OK(
+          AddSubstraitCallToArrow({kSubstraitLogarithmicFunctionsUri, function_name},
+                                  DecodeOptionlessOverflowableArithmetic(function_name)));
```
The logrithmic functions are not technically overflowable either.

drabastomek replied:
I changed the DecodeOptionlessUncheckedArithmetic and updated the tests.

westonpace reviewed (approved):
The Substrait options handling is off here, but not more off than any of the other functions, since Substrait's function definitions have changed since the original mappings were added.

 * Some functions, like `sign`, have no options.  You are currently expecting an overflow.
 * Floating point implementations do not have an overflow option.
 * Floating point implementations do usually have options for domain error and rounding mode

Rather than fix that here I have created https://issues.apache.org/jira/browse/ARROW-18133 .

I might see if someone else more familiar with the kernels can take a quick look at your `Exp` implementation just to make sure we aren't missing something.

westonpace commented:
Pinged @lidavidm / @pitrou for a brief review of the Exp kernel.  You can probably ignore the Substrait part if you want.

lidavidm commented on cpp/src/arrow/compute/kernels/base_arithmetic_internal.h:
```diff
@@ -485,6 +485,22 @@ struct NegateChecked {
+  static constexpr enable_if_decimal_value<Arg, T> Call(KernelContext*, Arg exp,
+                                                        Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    return exp.Exp();
```
Does this actually exist? I don't see a definition for it in (Basic)Decimal128

lidavidm replied:
I guess it's tested below so it should clearly exist…

lidavidm replied:
Ah, no, this isn't being used at all - decimals are being converted to floats first. Can we remove this?

drabastomek replied:
QQ: David, are you asking if the `enable_if_decimal_value` exists? I'm not sure I understand your comment, sorry...

pitrou replied:
@drabastomek {{Decimal128::Exp}} doesn't exist, so this definition wouldn't compile.

drabastomek replied:
Ah... I see. Ok, but the Substrait still can call Acero with a decimal value. I understand it'd be converted to float first to do the `Exp` but would we rather error on this or keep it? I'm happy to remove this and the relevant test just thinking out loud.

lidavidm replied:
The code here is dead code. We should remove it.

It doesn't matter what Substrait does; we currently implement most of these kernels by casting to float first. If that's not OK for Substrait, then we'll have to come back to implement actual decimal versions of these kernels.

westonpace replied:
FWIW, we should not be allowing this kind of implicit cast and should instead reject a plan containing a decimal exp call.  If the producer wants to allow the cast to float they can allow it.

However, this issue is beyond this PR.  I've created https://issues.apache.org/jira/browse/ARROW-18193

lidavidm reviewed (commented):
We should add the new function to the C++/Python docs

pitrou commented on cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:
```diff
@@ -1519,6 +1519,65 @@ TEST_F(TestUnaryArithmeticDecimal, AbsoluteValue) {
+  this->AssertUnaryOp(exp, "[-Inf, Inf]", "[0, Inf]");
+  // Arrays with NaNs
+  this->SetNansEqual(true);
+  this->AssertUnaryOp(exp, "[NaN]", "[NaN]");
+  this->AssertUnaryOp(exp, "[NaN]", "[NaN]");
+  this->AssertUnaryOp(exp, "[NaN]", "[NaN]");
```
Why three times?

drabastomek replied:
Good catch! Thanks @pitrou

drabastomek replied:
Updated the docs.

pitrou reviewed (changes requested):
Thanks @drabastomek . Can you also update https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst#arithmetic-functions ?

rok commented:
@drabastomek can I help move this forward?

westonpace commented:
@rok Yes, I think that would be helpful.  I had agreed with @drabastomek offline to finish this up for him but hadn't gotten around to it yet.  If you have a few cycles to address the review comments that would be appreciated!

drabastomek commented:
@westonpace @rok Thanks guys! I had this on my radar for this week. I addressed the comments but would be great if @rok could have another looks -- much appreciated!

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -450,6 +450,18 @@ Mixed time resolution temporal inputs will be cast to finest input resolution.
 +------------------+--------+------------------+----------------------+-------+
 | divide_checked   | Binary | Numeric/Temporal | Numeric/Temporal     | \(1)  |
 +------------------+--------+------------------+----------------------+-------+
+| exp              | Unary  | Numeric          | Numeric              |       |
```
We should be more precise when listing input and output types (see examples under "Logaritmic Functions" below).
Also you'll need to correct the table formatting after this :-)
```suggestion
| exp              | Unary  | Float32/Float64/Decimal | Float32/Float64      |       |
```

rok replied:
Wouldn't integer be a valid input type and hence input type remains `Numeric`? `Float32/Float64` as output makes sense.

This is out of scope for this PR, but how about `complex` values ([example](https://numpy.org/devdocs/reference/generated/numpy.exp.html))? Should Acero kernels support complex? What about Substrait?

rok replied:
I went with your suggestion, but I think we should probably still specify integers?

pitrou replied:
I don't know, does `exp` actually accept integers?

rok replied:
As per [cppreference](https://en.cppreference.com/w/cpp/numeric/math/exp) it does since c++11.

rok replied:
I'll admit it's not a terribly interesting case but still.

pitrou replied:
I'm not asking whether C++ supports it, but whether Arrow wants to support it :-)

As a comparison, our logarithm functions only support reals (either floating-point or decimals), not integers.

rok replied:
Integer support for logarithms feels more important than here. But I suppose implicit casting will cover the gap if needed? Shall I remove the integer support then?

rok replied:
Oh we're using `MakeUnaryArithmeticFunctionFloatingPoint` we can't remove it anyway.

rok reviewed (commented):
After addressing @pitrou's comments this looks pretty good to me.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -610,6 +610,15 @@ Result<Datum> Power(const Datum& left, const Datum& right,
+/// \brief Raise Euler's number to the specified power, element-wise.
+/// If the exponent value is null the result will be null.
+///
+/// \param[in] arg the base
```
I've reworded this a little.

rok commented on cpp/src/arrow/compute/kernels/base_arithmetic_internal.h:
```diff
@@ -485,6 +485,22 @@ struct NegateChecked {
 };
 
+struct Exp {
+  template <typename T, typename Arg>
```
Removed the decimal branch since `exp.Exp()` doesn't exists.

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -450,6 +450,18 @@ Mixed time resolution temporal inputs will be cast to finest input resolution.
 +------------------+--------+------------------+----------------------+-------+
+| exp              | Unary  | Numeric          | Numeric              |       |
++------------------+--------+------------------+----------------------+-------+
+| ln               | Unary  | Numeric          | Numeric              |       |
```
Removed.

rok commented on cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:
```diff
@@ -2224,6 +2229,10 @@ void RegisterScalarArithmetic(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(power_checked)));
 
+  // ----------------------------------------------------------------------
+  auto exp = MakeUnaryArithmeticFunctionFloatingPoint<Exp>("exp", exp_doc);
```
Perhaps `AddDecimalUnaryKernels<Exp>(exp.get());` could be added here? But since `Decimal128::Exp` doesn't exist we want implicit cast and keep `MakeUnaryArithmeticFunctionFloatingPoint`?

rok commented:
@pitrou I've agreed with @drabastomek to implement the review feedback. Could you please take another look?

pitrou commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -313,7 +330,18 @@ TEST(FunctionMapping, ValidCases) {
        "abcdef",
-       utf8()}};
+       utf8()},
+      {{kSubstraitLogarithmicFunctionsUri, "ln"}, {"1"}, {int8()}, "0", float64()},
+      {{kSubstraitLogarithmicFunctionsUri, "log10"}, {"1"}, {int8()}, "0", float64()},
+      {{kSubstraitLogarithmicFunctionsUri, "log2"}, {"2"}, {int8()}, "1", float64()},
+      {{kSubstraitLogarithmicFunctionsUri, "log1p"}, {"0"}, {int8()}, "0", float64()},
+      {{kSubstraitLogarithmicFunctionsUri, "logb"},
```
I'll note these tests don't have a lot of discriminating power.

rok replied:
Agreed. I've changed the `exp(0)`-like cases to be more discriminating.

Also note the comment above:
```
// These are not meant to be an exhaustive test of Substrait
// conformance.  Instead, we should test just enough to ensure
// we are mapping to the correct function
TEST(FunctionMapping, ValidCases) {
```

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -435,43 +435,45 @@ floating-point arguments will cast all arguments to floating-point, while mixed
++------------------+--------+-------------------------+----------------------+-------+
+| divide_checked   | Binary | Numeric/Temporal        | Numeric/Temporal     | \(1)  |
++------------------+--------+-------------------------+----------------------+-------+
+| exp              | Unary  | Float32/Float64/Decimal | Float32/Float64      |       |
```
```suggestion
| exp              | Unary  | Numeric                 | Float32/Float64      |       |
```

rok commented:
@pitrou do you think this needs another pass?

pitrou reviewed (approved):
LGTM now, thank you @rok

rok commented:
Nice! Thanks @drabastomek & all!

drabastomek commented:
Thank you all for the support with this!
